File: server/functions/src/models/domain/game/game_event.ts
import { Entity } from '../../../core/domain/entity';

export interface GameEvent<EventData = any> {
  readonly id: GameEventEntity.Id;
  readonly name: string;
  readonly description: string;
  readonly image?: string | null;
  readonly type: GameEventEntity.Type;
  readonly data: EventData;
}

export namespace GameEventEntity {
  export type Id = string;
  export type Type = string;

  export const validate = (gameEvent: any) => {
    const entity = Entity.createEntityValidator<GameEvent>(gameEvent, 'Game Event');

    entity.hasValue('id');
    entity.hasValue('name');
    entity.hasValue('description');
    entity.hasValue('type');
  };
}

File: client/lib/models/domain/game/game_event/game_event.dart
import 'package:flutter/foundation.dart';
import 'package:cash_flow/models/domain/game/game_event/game_event_type.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'game_event.g.dart';

abstract class GameEventData {
  Map<String, dynamic> toJson();
}

@JsonSerializable(createFactory: false)
class GameEvent {
  GameEvent({
    @required this.id,
    @required this.name,
    @required this.description,
    this.image,
    @required this.type,
    @required this.data,
  });

  factory GameEvent.fromJson(Map<String, dynamic> json) {
    final type = GameEventType.fromJson(json['type']);
    final eventData = type.parseGameEventData(json['data']);

    return GameEvent(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      image: json['image'],
      type: type,
      data: eventData,
    );
  }

  final String id;
  final String name;
  final String description;
  final String image;
  @JsonKey(fromJson: GameEventType.fromJson, toJson: GameEventType.toJson)
  final GameEventType type;
  final GameEventData data;

  Map<String, dynamic> toJson() => _$GameEventToJson(this);
}
